contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+
+It is not a sustainable solution to expect every vendor to add their vendor
+specific code inside Kubernetes to make their devices usable.
+Instead, we want a solution for vendors to be able to advertise their resources
+to Kubelet and monitor them without writing custom Kubernetes code.
+We also want to provide a consistent and portable solution for users to
+consume hardware devices across k8s clusters.
+
+This document describes a vendor independant solution to:
+  * Discovering and representing external devices
+  * Making these devices available to the containers using these devices and
+    cleaning them up afterwards
+  * Health Check of these devices
+
+Because devices are vendor dependant and have their own sets of problems
+and mechanisms, the solution we describe is a plugin mechanism that may run
+in a container deployed through the DaemonSets mechanism.
+The targeted devices include GPUs, High-performance NICs, FPGAs, InfiniBand,
+Storage devices, and other similar computing resources that require vendor
+specific initialization and setup.
+
+The goal is for a user to be able to enable vendor devices (e.g: GPUs) through
+the following simple steps:
+  * `kubectl create -f http://vendor.com/device-plugin-daemonset.yaml`
+  * When launching `kubectl describe nodes`, the devices appear in the node spec
+  * In the long term users will be able to select them through Resource Class
+
+# Use Cases
+
+ * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+   in my pod.
+ * I should be able to use that device without writing custom Kubernetes code.
+ * I want a consistent and portable solution to consume hardware devices
+   across k8s clusters.
+
+# Objectives
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through a pluggable mechanism.
+    * Which allows discovery and monitoring of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Advanced scheduling and resource selection (solved through
+   [#782](https://github.com/Kubernetes/community/pull/782)).
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# Proposed Implementation 1
+
+## TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
+document and once the device plugin makes itself known to kubelet, kubelet
+will interact with the device through three simple functions:
+  1. A `ListDevices` function for the kubelet to Discover the devices and
+     their properties.
+  2. An `Allocate` function which is called before container creation
+  3. A `HealthCheck` function to notify Kubelet whenever a device becomes
```

Comment:
Actually, wow about using `ListDevicesPeriodically` instead which just keeps returning a list of devices with states. I think this might be easier to work against than Watch because it is difficult to model devices disappearing in Watch.